src/asInput/index.jsx

```diff
@@ -51,22 +55,36 @@ const asInput = (WrappedComponent, inputType = undefined, labelFirst = true) =>
       this.handleBlur = this.handleBlur.bind(this);
 
       const id = this.props.id ? this.props.id : newId('asInput');
+      const isValid = this.props.validator ? true : this.props.isValid;
+      const validationMessage = this.props.validator ? '' : this.props.validationMessage;
       this.state = {
         id,
         value: this.props.value,
-        isValid: true,
+        isValid,
+        validationMessage,
         describedBy: [],
         errorId: `error-${id}`,
         descriptionId: `description-${id}`,
       };
     }
 
     componentWillReceiveProps(nextProps) {
+      const updatedState = {};
       if (nextProps.value !== this.props.value) {
-        this.setState({
-          value: nextProps.value,
-        });
+        updatedState.value = nextProps.value;
       }
+      if (nextProps.isValid !== this.props.isValid && !nextProps.validator) {
```

Comment:
All of the next three conditionals are dependent on `!nextProps.validator` - might this be simplified by adding a parent `if (!nextProps.validator)`?

Reply:
In my head, I liked this better than nested conditionals, but I could be persuaded if you feel strongly.

Reply:
nope, I don't feel strongly!